feat(revenue-analytics): Virtual revenue properties

rafaeelaudibert opened the pull request:
We want people to have access to revenue data on the `persons` table by doing something like `persons.$virt_revenue`. This PR achieves this by creating a new lazy table called `raw_revenue_analytics` that we can then join on. This implies people can get access to it either via `persons.revenue_analytics.revenue` OR `persons.$virt_revenue`. These can be used for filtering too.

People need to choose how we can find Stripe customers in PostHog - either via the `id`, the `email` or an arbitrary field on the customer `metadata` in Stripe.

Closes #33505

rafaeelaudibert committed:
refactor: Mark revenue analytics properties as virtual

This simplifies our handling in the UI, so let's do this

rafaeelaudibert committed:
feat: Initial work for virtual `revenue` fields for `persons`

We want people to have access to revenue data on the `persons` table by doing something like `persons.$virt_revenue`. This PRs achieves this by creating a new lazy table called `raw_revenue_analytics` that we can then join on. We're currently hardcoding the join on the Stripe ID which looks like `cus_xxxx`. This will only work if people are aliasing their PostHog persons to the ID they have in Stripe - which is unlikely. We'll make this configurable in the next commit.

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
feat: Add configurable persons join for revenue analytics

People can now choose how we'll join between our `persons` table and the `customer` view coming from Revenue Analytics

Follow-up commit will implement the join

rafaeelaudibert committed:
feat: Consider persons mode for revenue analytics

We're now considering the person mode when querying for the Revenue Analytics virtual fields. There's a bug happening when a person has more than one distinct id, will need some help investigating it

rafaeelaudibert committed:
fix: Fix mypy types

rafaeelaudibert committed:
fix: Remove breakpoint

rafaeelaudibert commented on posthog/hogql/database/schema/test/test_revenue_analytics.py:
```diff
@@ -0,0 +1,273 @@
+            # 1. I can only get the revenue to work if I select * first, doesnt work if select just revenue
+            # 2. This is returning an entry for each of the distinct IDs which is understandable but could probably get the one that exists only
+            # 3. If I change the select to a specific person `where id = {}` then it fails to return for the matching email, never returns revenue
+            assert response.results[0][6] == Decimal("429.7424")
```
@mariusandra This is the test where I can reproduce the problem, it only happens when I have a person with more than one distinct_id. The easiest way to debug this is by adding a breakpoint here and running queries via `execute_hogql_query`. Make sure you include `modifiers=self.MODIFIERS` to be able to parse the mocked data warehouse tables appropriately

mariusandra replied:
I'm unable to run the tests in my debugger in pycharm for what looks like unrelated reasons.

```
AttributeError: module 'django.conf.global_settings' has no attribute 'CLICKHOUSE_CONN_POOL_MAX'
```

Can you tell me what query does this generate? Let's start by making sure it tries to do the right thing before debugging. So... what does it generate, and what should it generate?

robbie-c commented on posthog/api/team.py:
```diff
@@ -176,10 +176,21 @@ class TeamRevenueAnalyticsConfigSerializer(serializers.ModelSerializer):
     events = serializers.JSONField(required=False)
     goals = serializers.JSONField(required=False)
     filter_test_accounts = serializers.BooleanField(required=False)
+    persons_join_mode = serializers.ChoiceField(
```
It might be more appropriate to have this as a HogQLQueryModifier option, so that changing it will cache-bust properly

mariusandra replied:
The naming of this is very confusing. I immediately think of "persons on events", not "revenue analytics". Please make it more verbose.

I'm also not a fan of dumping everything on the team model, but 🤷

rafaeelaudibert commented on frontend/src/taxonomy/core-filter-definitions-by-group.json:
```diff
@@ -3591,6 +3591,18 @@
+            "type": "Numeric",
+            "virtual": true
+        },
+        "$virt_revenue_last_30_days": {
```
Yes, they might, and in that case that will give us the itch to make these custom 😆

rafaeelaudibert commented on products/revenue_analytics/frontend/settings/PersonsJoinModeConfiguration.tsx:
```diff
@@ -0,0 +1,86 @@
+                    onChange={updatePersonsJoinMode}
+                    options={[
+                        {
+                            value: RevenueAnalyticsPersonsJoinMode.ID,
```
I just don't think it's that much work to support what I've built here, tbh

robbie-c replied:
All good 😄

rafaeelaudibert committed:
fix: Make $virt_revenue fields work

The basic problem we had is that we were exposing the revenue analytics table inside `persons` but that doesnt make sense because we need to connect revenue with a specific `distinct_id` instead, therefore it should be part of the `distinct_id` table, which we've done now.

There's a small problem with our approach, currently: we can properly display this field as a table, but it fails when trying to filter because we have a WINDOW function on a field that goes to the `where` which isnt supported - weird, but ok.

We need to find a better way to expose this field in the parent `persons` table such that we only have a *single* value in it without using any kind of aggregation function that would then fail for two-fold: not using group by, not being able to use window on WHERE.

Subquery comes to mind, but it's tricky to get them right.

rafaeelaudibert committed:
refactor: Use modifiers for revenue analytics person join

Rather than storing this on the database, we can store this as modifiers, this makes the code much simpler, removes the need for a migration, etc.

rafaeelaudibert commented on posthog/hogql/database/database.py:
```diff
@@ -811,6 +795,52 @@ def define_mappings(store: TableStore, get_table: Callable):
             except Exception as e:
                 capture_exception(e)
 
+    # with timings.measure("revenue_analytics_tables"):
+    #     database.persons.fields["pdi"] = LazyJoin(
+    #         from_field=["pdi", "distinct_id"],
+    #         join_table=database.raw_revenue_analytics,
+    #         join_function=join_with_revenue_analytics_table,
+    #     )
```
TODO: Remove this

rafaeelaudibert commented on posthog/hogql/database/database.py:
```diff
@@ -811,6 +795,52 @@ def define_mappings(store: TableStore, get_table: Callable):
+                timings=timings,
+                properties_path=["poe", "properties"],
+            )
+        with timings.measure("revenue"):
+            field_name = "$virt_revenue"
+            database.persons.fields[field_name] = ast.ExpressionField(
+                name=field_name, expr=parse_expr("sum(pdi.revenue_analytics.revenue) OVER(PARTITION BY persons.id)")
+            )
+            poe.fields[field_name] = ast.ExpressionField(
+                name=field_name, expr=parse_expr("sum(pdi.revenue_analytics.revenue) OVER(PARTITION BY persons.id)")
+            )
```
@mariusandra @robbie-c Ok, this is now almost working.

Basically, because `distinct_id` is something that we find on the `pdi` table, I added the new `revenue_analytics` table as a lazy table on THAT table rather than `persons`, and this works very well for each individual `distinct_id`! The part I thought was going to be hard (matching) is now solved, and we now have the easy part to solve - ha!

The problem here is the following: we wanna have a single value for this virtual field when querying `persons`, so it makes sense that we need to "group" the values from all distinct IDs. The problem is: how do I group them?

1. I can't do GROUP BY because it would require me to group by all the other properties I'm using on an arbitrary query, and that doesn't sound great.
2. Using a window function like I'm doing here is promising - you can see tests are passing - but if I try filtering by this (for example, revenue > 1000), then it fails because you can't include a window function on a WHERE clause.
3. Subquery is the way to go, I guess? I couldnt get it to work because Clickhouse SQL works differently to normal SQL and I can't refer to `persons.id` inside the subquery

Any suggestions/clues/desire to try and make this work for me? I'm pretty sure that we can make virtual fields work in general if we can figure out a way to make the revenue ones work

mariusandra replied:
I didn't check the code, but yes, a subquery is what you want... 

I'm not sure how trivial any of that will be, but effectively you want to translate from

```sql
select a,b,c,d,e from persons
```

to

```sql
select x.a as a, x.b as b, x.c as c, x.d as d, sum(x.e) as e from (
   select a,b,c,d,e from persons
) as x
```

through some AST transformations somewhat automatically... 

But I wonder if this is getting too complex for our own good? Perhaps you can do some grouping on the other side of the join? So the table you're joining already has "person_id" it? So `left join ( select person_id, sum(money) as moolah from (select person_id, distinct_id, ... ) group by person_id )`

I don't know how useful the answer above is, as I again didn't go deep, but perhaps it can help somehow 🙃

rafaeelaudibert replied:
> But I wonder if this is getting too complex for our own good?

For my own good, absolutely. 😆 

> Perhaps you can do some grouping on the other side of the join? So the table you're joining already has "person_id" it?

Hmmm, that's an option, this way I can join from `persons` directly and there's no grouping needed, the other table can handle summing it all up. Cool, that might be the way forward. I'm coming back to this in 1 week, will try that approach, thank you!

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Fix mypy types

rafaeelaudibert committed:
feat: Reimplement `revenue_analytics` table using `person_id`

Rather than implementing this table at the `pdi` level with the `distinct_id`, let's instead do it at the `persons` level. This sounded tricky at first, but we can easily achieve this by exposing the `person_id` from the `customer` view in Revenue Analytics. On the revenue events, also expose that and simply join on that field, group by it, sum the values and boom, easy!

Thanks to @mariusandra for the idea, worked like a charm

rafaeelaudibert committed:
refactor: Extract virtual fields to separate function

rafaeelaudibert committed:
refactor: Simplify values check

Dont fill fields for virtual properties

rafaeelaudibert committed:
tests: Fix tests after changes to modifiers

We have a new modifier, which means the caches changed because of the new key

rafaeelaudibert commented:
@mariusandra @robbie-c I have a working solution heavily inspired by Marius' recent suggestions.

There's a lot of UI changes to support the new HogQL modifier, but the backend is actually very simple :)

rafaeelaudibert commented on posthog/hogql/database/schema/util/where_clause_extractor.py:
```diff
@@ -445,6 +445,11 @@ def visit_tuple(self, node: ast.Tuple) -> bool:
     def visit_array(self, node: ast.Tuple) -> bool:
         return all(self.visit(arg) for arg in node.exprs)
```
8 months old code, not me

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Fix incorrect test values

After merging with master we're missing some revenue entries, so let's fix the numbers here

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

robbie-c commented on posthog/hogql/database/schema/test/__snapshots__/test_revenue_analytics.ambr:
```diff
@@ -0,0 +1,974 @@
+                  toString(if(not(empty(events__override.distinct_id)), events__override.person_id, events.person_id)) AS customer_id,
+                  NULL AS invoice_id,
+                  NULL AS subscription_id,
+                  toString(events.`$session_id`) AS session_id,
```
I tested that clickhouse is pruning these unused columns correctly with some help from Claude. (TLDR it is)

* EXPLAIN PIPELINE header=1 $THIS_QUERY
* cmd+f session_id -> it's not there

You might want to run this manually with / without those lines deleted and check the bytes read, just to be sure

robbie-c commented on posthog/hogql/database/schema/persons_pdi.py:
```diff
@@ -85,11 +91,7 @@ def persons_pdi_join(
-        "distinct_id": StringDatabaseField(name="distinct_id", nullable=False),
-        "person_id": StringDatabaseField(name="person_id", nullable=False),
-    }
+    fields: dict[str, FieldOrTable] = FIELDS
```
Hmm, I'm not sure about this

We mutate several tables' fields during `create_hogql_database`

If every HogQL DB uses a shared field object, it will pick up changes from another `create_hogql_database` call with different modifiers. This would be a nightmare to debug, and more likely to show up in prod than in testing.

rafaeelaudibert committed:
fix: Avoid using shared object

rafaeelaudibert committed:
tests: Add insight trends test

rafaeelaudibert committed:
refactor: Python formatting

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

rafaeelaudibert commented:
Im breaking this PR down into smaller PRs, this should make reviewing this PR easier eventually
